### PR TITLE
lifecycle: move PartsLifecycle to LifecycleService

### PR DESCRIPTION
**Pull request by lengau**
- [x] Have you followed the guidelines for contributing?
- [x] Have you signed the [CLA](http://www.ubuntu.com/legal/contributors/)?
- [x] Have you successfully run `tox`?

-----

The business logic for various commands will live in service classes, so the business logic for lifecycle should live there too.

### CODE REVIEW

File: craft_application/services/lifecycle.py [hunk 1]
```diff
@@ -110,6 +111,7 @@ def __init__(
         cache_dir: Path | str,
         **lifecycle_kwargs: Any,
     ) -> None:
+        super().__init__(app, project)
         self._app = app
```
Comment (tigarmo) [inline craft_application/services/lifecycle.py:115]: I guess these two setters aren't needed anymore right?

File: craft_application/services/lifecycle.py [hunk 2]
```diff
@@ -110,6 +111,7 @@ def __init__(
         cache_dir: Path | str,
```
Comment (tigarmo) [inline craft_application/services/lifecycle.py:111]: unrelated to this PR but this alternative to Optional needs Python >=3.10 right? Is that an issue?

Comment (lengau) [reply craft_application/services/lifecycle.py:111]: `from __future__ import annotations` brings it in. I'm less thrilled personally about changing `Optional[type]` to `type | None` than about pipes for union types and things like `list[str]`, but once you enable that mypy and complain about using `Optional`. Worth bringing up to the team what we'd like our standard to be going forward (assuming we can tell mypy/pyright to allow `Optional` if we choose).

File: craft_application/services/lifecycle.py (hunk 1 above)
Comment (lengau) [reply craft_application/services/lifecycle.py]: Good catch!

**Review by tigarmo — APPROVED** [review]

**Pull request merged**